crates/matrix-crypto-ffi/src/lib.rs

```diff
@@ -23,18 +23,308 @@ pub use device::Device;
+
+    let mut sessions = Vec::new();
+
+    for session_pickle in data.sessions {
+        let pickle =
+            Session::from_libolm_pickle(&session_pickle.pickle, &data.pickle_key)?.pickle();
+
+        let creation_time = SecondsSinceUnixEpoch(UInt::from_str(&session_pickle.creation_time)?);
+        let last_use_time = SecondsSinceUnixEpoch(UInt::from_str(&session_pickle.last_use_time)?);
+
+        let pickle = matrix_sdk_crypto::olm::PickledSession {
+            pickle,
+            sender_key: Curve25519PublicKey::from_base64(&session_pickle.sender_key)?,
+            created_using_fallback_key: session_pickle.created_using_fallback_key,
+            creation_time,
+            last_use_time,
+        };
+
+        let session = matrix_sdk_crypto::olm::Session::from_pickle(
+            user_id.clone(),
+            device_id.clone(),
+            identity_keys.clone(),
+            pickle,
+        );
+
+        sessions.push(session);
+        processed_steps += 1;
+        listener(processed_steps, total_steps);
+    }
+
+    let mut inbound_group_sessions = Vec::new();
+
+    for session in data.inbound_group_sessions {
+        let pickle =
+            InboundGroupSession::from_libolm_pickle(&session.pickle, &data.pickle_key)?.pickle();
+
+        let pickle = matrix_sdk_crypto::olm::PickledInboundGroupSession {
+            pickle,
+            sender_key: session.sender_key,
+            signing_key: session
+                .signing_key
+                .into_iter()
+                .map(|(k, v)| Ok((DeviceKeyAlgorithm::try_from(k)?, v)))
+                .collect::<Result<_>>()?,
+            room_id: RoomId::parse(session.room_id)?,
+            forwarding_chains: session.forwarding_chains,
+            imported: session.imported,
+            backed_up: session.backed_up,
+            history_visibility: None,
+        };
+
+        let session = matrix_sdk_crypto::olm::InboundGroupSession::from_pickle(pickle)?;
+
+        inbound_group_sessions.push(session);
+        processed_steps += 1;
+        listener(processed_steps, total_steps);
+    }
+
+    let recovery_key =
+        data.backup_recovery_key.map(|k| RecoveryKey::from_base58(k.as_str())).transpose()?;
+
+    let cross_signing = PrivateCrossSigningIdentity::empty((*user_id).into());
+    runtime.block_on(cross_signing.import_secrets_unchecked(
+        data.cross_signing.master_key.as_deref(),
+        data.cross_signing.self_signing_key.as_deref(),
+        data.cross_signing.user_signing_key.as_deref(),
+    ))?;
+
+    data.cross_signing.master_key.zeroize();
+    data.cross_signing.self_signing_key.zeroize();
+    data.cross_signing.user_signing_key.zeroize();
+
+    processed_steps += 1;
+    listener(processed_steps, total_steps);
+
+    let tracked_users = data
+        .tracked_users
+        .into_iter()
+        .map(|u| Ok(((parse_user_id(&u)?), true)))
+        .collect::<Result<Vec<(Box<UserId>, bool)>>>()?;
```

Comment:
Type inference wise, specifying the type inside the `Vec` should not be needed. Also I'd prefer `anyhow::Result` being qualified and the moving as much information out of the turbofish as possible for readability of the `collect` line, but I wouldn't insist on it.

```suggestion
    let tracked_users: Vec<_> = data
        .tracked_users
        .into_iter()
        .map(|u| Ok(((parse_user_id(&u)?), true)))
        .collect::<anyhow::Result<_>>()?;
```

Reply:
Also could use `UserId::parse` here since the separate helper is only for converting the error to `CryptoError`, right?

Reply:
`parse_user_id` adds a better error message, since we had a problem of having a user id in the wild that failed the parse step. The error returned by Ruma wasn't telling us much, so as a workaround I added this in the hope that we would get another report quickly.

Reply:
I see. Maybe we should include the thing that failed to parse in the Ruma error type 🤔

Reply:
That would probably help, but then again it would probably leak user ids in error logs. :shrug: